strawberry/fastapi/router.py

```diff
@@ -38,36 +45,34 @@ async def __get_root_value():
 
     @staticmethod
     def __get_context_getter(
-        custom_getter: Callable[..., Optional[Union[BaseContext, Dict[str, Any]]]]
-    ) -> Callable[..., Union[BaseContext, Dict[str, Any]]]:
+        custom_getter: Callable[..., Optional[CustomContext]]
+    ) -> Callable[..., CustomContext]:
         def dependency(
-            custom_getter: Optional[Union[BaseContext, Dict[str, Any]]],
+            custom_getter: Optional[CustomContext],
             background_tasks: BackgroundTasks,
             request: Request = None,
             response: Response = None,
             ws: WebSocket = None,
-        ) -> Union[
-            BaseContext, Dict[str, Union[Any, BackgroundTasks, Request, WebSocket]]
-        ]:
+        ) -> MergedContext:
+            default_dict = {
+                "request": request or ws,
+                "background_tasks": background_tasks,
+                "response": response,
+            }
             if isinstance(custom_getter, BaseContext):
                 custom_getter.request = request or ws
                 custom_getter.background_tasks = background_tasks
+                custom_getter.response = response
                 return custom_getter
             elif isinstance(custom_getter, dict):
                 return {
-                    "request": request or ws,
-                    "background_tasks": background_tasks,
-                    **(custom_getter or {}),
-                }
-            elif custom_getter == None:
-                return {
-                    "request": request or ws,
-                    "background_tasks": background_tasks,
+                    **default_dict,
+                    **custom_getter,
                 }
+            elif custom_getter is None:
+                return default_dict
             else:
-                raise TypeError(
-                    "The custom context dependency is neither a class that inherits from BaseContext nor a dictionary"
-                )
+                raise InvalidCustomContext()
```

Comment:
Should we pass the invalid context to this exception so that it can use it in the exception string?

Reply:
It would require `strawberry.exceptions` to import `BaseContext` from `strawberry.fastapi` to correctly type the argument to the exception's `__init__` method. Might that pose a problem of cross-dependency? Perhaps we make an `exceptions.py` file in `strawberry.fastapi`?